File: python/misah_ref/plane_dem.py
"""Reference implementation of the plane-DEM intersection kernel.

This is the numpy oracle against which the Rust kernel in ``src/kernels/plane_dem.rs``
is validated, and the fallback used when the compiled extension is unavailable
(QGIS plugins cannot rely on binary wheels being installable).

The intersection is computed as the zero-level set of the signed distance

    f(x, y) = n . (X - P0),    X = (x, y, z_dem(x, y))

sampled at the DEM nodes, extracted by marching squares. Working on the signed
distance rather than on a z = z(x, y) plane expression keeps vertical planes
(where the plane has no single-valued z) as an ordinary case.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

import numpy as np

# Marching-squares edge crossings per cell configuration. Corners are numbered
# 0=top-left, 1=top-right, 2=bottom-right, 3=bottom-left; edges are named after
# the corner pair they join. Case index has bit i set when corner i is positive.
_EDGE_CORNERS = {
    "top": (0, 1),
    "right": (1, 2),
    "bottom": (3, 2),
    "left": (0, 3),
}

_CASE_EDGES = {
    0: (),
    1: (("left", "top"),),
    2: (("top", "right"),),
    3: (("left", "right"),),
    4: (("right", "bottom"),),
    5: (("left", "top"), ("right", "bottom")),      # ambiguous: resolved on centre
    6: (("top", "bottom"),),
    7: (("left", "bottom"),),
    8: (("left", "bottom"),),
    9: (("top", "bottom"),),
    10: (("left", "bottom"), ("top", "right")),     # ambiguous: resolved on centre
    11: (("right", "bottom"),),
    12: (("left", "right"),),
    13: (("top", "right"),),
    14: (("left", "top"),),
    15: (),
}


@dataclass(frozen=True)
class GeoTransform:
    """GDAL-style affine transform, mapping (col, row) offsets to (x, y)."""

    x_origin: float
    pixel_width: float
    row_rotation: float
    y_origin: float
    col_rotation: float
    pixel_height: float

    @classmethod
    def from_esri_ascii_header(
        cls, xllcorner: float, yllcorner: float, cellsize: float, nrows: int
    ) -> "GeoTransform":
        return cls(
            x_origin=xllcorner,
            pixel_width=cellsize,
            row_rotation=0.0,
            y_origin=yllcorner + nrows * cellsize,
            col_rotation=0.0,
            pixel_height=-cellsize,
        )

    def node_coords(self, nrows: int, ncols: int) -> Tuple[np.ndarray, np.ndarray]:
        """Return (x, y) arrays of shape (nrows, ncols) for the DEM node centres.

        DEM values are point samples at cell centres, hence the +0.5 offsets.
        """
        rows, cols = np.meshgrid(
            np.arange(nrows, dtype=float) + 0.5,
            np.arange(ncols, dtype=float) + 0.5,
            indexing="ij",
        )
        x = self.x_origin + cols * self.pixel_width + rows * self.row_rotation
        y = self.y_origin + cols * self.col_rotation + rows * self.pixel_height
        return x, y


def plane_normal(dip_dir_degr: float, dip_angle_degr: float) -> np.ndarray:
    """Upward-pointing unit normal of a geological plane, in (East, North, Up).

    Sanity anchors: a horizontal plane (dip 0) gives (0, 0, 1); a vertical plane
    (dip 90) gives a horizontal normal pointing along the dip direction.
    """
    az = np.radians(dip_dir_degr)
    dip = np.radians(dip_angle_degr)
    return np.array(
        [np.sin(az) * np.sin(dip), np.cos(az) * np.sin(dip), np.cos(dip)],
        dtype=float,
    )


def signed_distance(
    dem: np.ndarray,
    geotransform: GeoTransform,
    src_pt: Tuple[float, float, float],
    dip_dir_degr: float,
    dip_angle_degr: float,
    nodata: Optional[float] = None,
) -> np.ndarray:
    """Signed distance from the plane to each DEM node, in ground units.

    NODATA nodes and non-finite elevations become NaN and are excluded downstream.
    """
    z = np.asarray(dem, dtype=float)
    if nodata is not None:
        z = np.where(np.isclose(z, nodata), np.nan, z)

    nrows, ncols = z.shape
    x, y = geotransform.node_coords(nrows, ncols)
    n = plane_normal(dip_dir_degr, dip_angle_degr)
    x0, y0, z0 = src_pt

    return n[0] * (x - x0) + n[1] * (y - y0) + n[2] * (z - z0)


def _interp_edge(fa, fb, xa, ya, za, xb, yb, zb):
    """Zero crossing along an edge, by linear interpolation on the signed distance."""
    t = fa / (fa - fb)
    return (xa + t * (xb - xa), ya + t * (yb - ya), za + t * (zb - za))


def intersect_plane_dem(
    dem: np.ndarray,
    geotransform: GeoTransform,
    src_pt: Tuple[float, float, float],
    dip_dir_degr: float,
    dip_angle_degr: float,
    nodata: Optional[float] = None,
) -> Tuple[np.ndarray, np.ndarray]:
    """Intersect an unbounded geological plane with a DEM.

    Returns ``(points, segments)`` where ``points`` is an (N, 3) array of
    intersection vertices and ``segments`` an (M, 2) array of indices into
    ``points``, one row per marching-squares chord. Segments let callers rebuild
    polylines; the bare vertices feed attitude inversion directly.
    """
    f = signed_distance(
        dem, geotransform, src_pt, dip_dir_degr, dip_angle_degr, nodata
    )
    nrows, ncols = f.shape
    x, y = geotransform.node_coords(nrows, ncols)
    z = np.asarray(dem, dtype=float)
    if nodata is not None:
        z = np.where(np.isclose(z, nodata), np.nan, z)

    points: list = []
    segments: list = []
    # Vertices are shared between adjacent cells; key them by edge identity so a
    # crossing is emitted once and segments stay topologically connected.
    vertex_of_edge: dict = {}

    def edge_vertex(r0, c0, r1, c1):
        key = ((r0, c0), (r1, c1))
        hit = vertex_of_edge.get(key)
        if hit is not None:
            return hit
        pt = _interp_edge(
            f[r0, c0], f[r1, c1],
            x[r0, c0], y[r0, c0], z[r0, c0],
            x[r1, c1], y[r1, c1], z[r1, c1],
        )
        idx = len(points)
        points.append(pt)
        vertex_of_edge[key] = idx
        return idx

    for r in range(nrows - 1):
        for c in range(ncols - 1):
            # corner order: 0=(r,c) 1=(r,c+1) 2=(r+1,c+1) 3=(r+1,c)
            corners = ((r, c), (r, c + 1), (r + 1, c + 1), (r + 1, c))
            vals = [f[i, j] for i, j in corners]
            if not all(np.isfinite(v) for v in vals):
                continue
            # A node exactly on the plane would make the sign test ill-defined;
            # nudging it off by an ulp-scale epsilon keeps the topology consistent.
            vals = [v if v != 0.0 else np.finfo(float).tiny for v in vals]

            case = sum((1 << i) for i, v in enumerate(vals) if v > 0.0)
            edges = _CASE_EDGES[case]
            if not edges:
                continue

            if case in (5, 10):
                # Saddle: the cell centre tells which of the two pairings is right.
                centre = sum(vals) / 4.0
                if (centre > 0.0) != (case == 5):
                    edges = (
                        (("left", "bottom"), ("top", "right"))
                        if case == 5
                        else (("left", "top"), ("right", "bottom"))
                    )

            for e0, e1 in edges:
                idx = []
                for edge in (e0, e1):
                    a, b = _EDGE_CORNERS[edge]
                    (r0, c0), (r1, c1) = corners[a], corners[b]
                    idx.append(edge_vertex(r0, c0, r1, c1))
                segments.append(tuple(idx))

    pts = np.array(points, dtype=float).reshape(-1, 3)
    segs = np.array(segments, dtype=np.int64).reshape(-1, 2)
    return pts, segs


def read_esri_ascii(path: str) -> Tuple[np.ndarray, GeoTransform, float]:
    """Minimal ESRI ASCII grid reader, sufficient for the test fixtures."""
    header = {}
    with open(path) as src:
        for _ in range(6):
            key, value = src.readline().split()
            header[key.upper()] = float(value)
        data = np.loadtxt(src)

    nrows, ncols = int(header["NROWS"]), int(header["NCOLS"])
    data = data.reshape(nrows, ncols)
    gt = GeoTransform.from_esri_ascii_header(
        header["XLLCORNER"], header["YLLCORNER"], header["CELLSIZE"], nrows
    )
    return data, gt, header["NODATA_VALUE"]

File: src/kernels/plane_dem.rs
//! Plane-DEM intersection.
//!
//! The intersection is the zero-level set of the signed distance
//!
//! ```text
//! f(x, y) = n . (X - P0),    X = (x, y, z_dem(x, y))
//! ```
//!
//! sampled at the DEM nodes and extracted by marching squares. Working on the
//! signed distance rather than on a `z = z(x, y)` plane expression keeps vertical
//! planes -- which have no single-valued z -- as an ordinary case.
//!
//! Mirrors `python/misah_ref/plane_dem.py`; `tests/test_plane_dem.py` validates
//! both against the geoSurfDEM golden dataset.

use std::collections::HashMap;

/// GDAL-style affine transform, mapping (col, row) offsets to (x, y).
#[derive(Clone, Copy, Debug)]
pub struct GeoTransform {
    pub x_origin: f64,
    pub pixel_width: f64,
    pub row_rotation: f64,
    pub y_origin: f64,
    pub col_rotation: f64,
    pub pixel_height: f64,
}

impl GeoTransform {
    /// DEM values are point samples at cell centres, hence the +0.5 offsets.
    #[inline]
    pub fn node_xy(&self, row: usize, col: usize) -> (f64, f64) {
        let (r, c) = (row as f64 + 0.5, col as f64 + 0.5);
        (
            self.x_origin + c * self.pixel_width + r * self.row_rotation,
            self.y_origin + c * self.col_rotation + r * self.pixel_height,
        )
    }
}

/// Upward-pointing unit normal of a geological plane, in (East, North, Up).
///
/// Sanity anchors: a horizontal plane (dip 0) gives (0, 0, 1); a vertical plane
/// (dip 90) gives a horizontal normal pointing along the dip direction.
pub fn plane_normal(dip_dir_degr: f64, dip_angle_degr: f64) -> [f64; 3] {
    let az = dip_dir_degr.to_radians();
    let dip = dip_angle_degr.to_radians();
    [az.sin() * dip.sin(), az.cos() * dip.sin(), dip.cos()]
}

/// Corner numbering: 0 = (r, c), 1 = (r, c+1), 2 = (r+1, c+1), 3 = (r+1, c).
/// Edges are named after the corner pair they join.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
enum Edge {
    Top,
    Right,
    Bottom,
    Left,
}

impl Edge {
    #[inline]
    fn corners(self) -> (usize, usize) {
        match self {
            Edge::Top => (0, 1),
            Edge::Right => (1, 2),
            Edge::Bottom => (3, 2),
            Edge::Left => (0, 3),
        }
    }
}

/// Marching-squares chords per case index; bit i is set when corner i is positive.
/// Complementary cases (i and 15-i) share their chords.
#[inline]
fn case_edges(case: u8) -> &'static [(Edge, Edge)] {
    match case {
        0 | 15 => &[],
        1 | 14 => &[(Edge::Left, Edge::Top)],
        2 | 13 => &[(Edge::Top, Edge::Right)],
        3 | 12 => &[(Edge::Left, Edge::Right)],
        4 | 11 => &[(Edge::Right, Edge::Bottom)],
        6 | 9 => &[(Edge::Top, Edge::Bottom)],
        7 | 8 => &[(Edge::Left, Edge::Bottom)],
        5 => &[(Edge::Left, Edge::Top), (Edge::Right, Edge::Bottom)],
        10 => &[(Edge::Left, Edge::Bottom), (Edge::Top, Edge::Right)],
        _ => unreachable!(),
    }
}

/// The two alternative pairings for a saddle cell.
#[inline]
fn saddle_alternative(case: u8) -> &'static [(Edge, Edge)] {
    match case {
        5 => &[(Edge::Left, Edge::Bottom), (Edge::Top, Edge::Right)],
        10 => &[(Edge::Left, Edge::Top), (Edge::Right, Edge::Bottom)],
        _ => unreachable!(),
    }
}

pub struct Intersection {
    /// Intersection vertices, as (x, y, z) triples.
    pub points: Vec<[f64; 3]>,
    /// Index pairs into `points`, one per marching-squares chord.
    pub segments: Vec<[usize; 2]>,
}

/// Intersect an unbounded geological plane with a DEM.
///
/// `dem` is row-major with `nrows * ncols` elements, rows running north to south.
/// Non-finite elevations, and those matching `nodata`, are excluded.
pub fn intersect_plane_dem(
    dem: &[f64],
    nrows: usize,
    ncols: usize,
    geotransform: &GeoTransform,
    src_pt: [f64; 3],
    dip_dir_degr: f64,
    dip_angle_degr: f64,
    nodata: Option<f64>,
) -> Intersection {
    assert_eq!(dem.len(), nrows * ncols, "DEM length does not match its shape");

    let normal = plane_normal(dip_dir_degr, dip_angle_degr);

    let elevation = |row: usize, col: usize| -> f64 {
        let z = dem[row * ncols + col];
        match nodata {
            Some(nd) if z == nd => f64::NAN,
            _ => z,
        }
    };

    let signed_distance = |row: usize, col: usize| -> f64 {
        let (x, y) = geotransform.node_xy(row, col);
        let z = elevation(row, col);
        normal[0] * (x - src_pt[0]) + normal[1] * (y - src_pt[1]) + normal[2] * (z - src_pt[2])
    };

    let mut points: Vec<[f64; 3]> = Vec::new();
    let mut segments: Vec<[usize; 2]> = Vec::new();
    // Vertices are shared between adjacent cells; key them by edge identity so a
    // crossing is emitted once and segments stay topologically connected.
    let mut vertex_of_edge: HashMap<[usize; 4], usize> = HashMap::new();

    for r in 0..nrows.saturating_sub(1) {
        for c in 0..ncols.saturating_sub(1) {
            let corners = [(r, c), (r, c + 1), (r + 1, c + 1), (r + 1, c)];
            let mut vals = [0.0f64; 4];
            let mut usable = true;
            for (i, &(cr, cc)) in corners.iter().enumerate() {
                let f = signed_distance(cr, cc);
                if !f.is_finite() {
                    usable = false;
                    break;
                }
                // A node exactly on the plane would make the sign test ill-defined;
                // nudging it off by an ulp-scale epsilon keeps the topology consistent.
                vals[i] = if f == 0.0 { f64::MIN_POSITIVE } else { f };
            }
            if !usable {
                continue;
            }

            let mut case = 0u8;
            for (i, &v) in vals.iter().enumerate() {
                if v > 0.0 {
                    case |= 1 << i;
                }
            }

            let mut chords = case_edges(case);
            if chords.is_empty() {
                continue;
            }
            if case == 5 || case == 10 {
                // Saddle: the cell centre tells which of the two pairings is right.
                let centre = vals.iter().sum::<f64>() / 4.0;
                if (centre > 0.0) != (case == 5) {
                    chords = saddle_alternative(case);
                }
            }

            for &(e0, e1) in chords {
                let mut ends = [0usize; 2];
                for (slot, edge) in [e0, e1].iter().enumerate() {
                    let (a, b) = edge.corners();
                    let (r0, c0) = corners[a];
                    let (r1, c1) = corners[b];
                    let key = [r0, c0, r1, c1];
                    ends[slot] = match vertex_of_edge.get(&key) {
                        Some(&idx) => idx,
                        None => {
                            let (xa, ya) = geotransform.node_xy(r0, c0);
                            let (xb, yb) = geotransform.node_xy(r1, c1);
                            let (za, zb) = (elevation(r0, c0), elevation(r1, c1));
                            let (fa, fb) = (signed_distance(r0, c0), signed_distance(r1, c1));
                            let t = fa / (fa - fb);
                            let idx = points.len();
                            points.push([
                                xa + t * (xb - xa),
                                ya + t * (yb - ya),
                                za + t * (zb - za),
                            ]);
                            vertex_of_edge.insert(key, idx);
                            idx
                        }
                    };
                }
                segments.push(ends);
            }
        }
    }

    Intersection { points, segments }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn flat_dem(nrows: usize, ncols: usize, z: f64) -> Vec<f64> {
        vec![z; nrows * ncols]
    }

    fn unit_transform() -> GeoTransform {
        GeoTransform {
            x_origin: 0.0,
            pixel_width: 1.0,
            row_rotation: 0.0,
            y_origin: 10.0,
            col_rotation: 0.0,
            pixel_height: -1.0,
        }
    }

    #[test]
    fn horizontal_plane_normal_points_up() {
        let n = plane_normal(0.0, 0.0);
        assert!((n[0]).abs() < 1e-12 && (n[1]).abs() < 1e-12 && (n[2] - 1.0).abs() < 1e-12);
    }

    #[test]
    fn vertical_plane_normal_is_horizontal() {
        let n = plane_normal(90.0, 90.0);
        assert!((n[0] - 1.0).abs() < 1e-12 && n[2].abs() < 1e-12);
    }

    #[test]
    fn flat_dem_below_plane_yields_nothing() {
        let dem = flat_dem(10, 10, 0.0);
        let out = intersect_plane_dem(
            &dem, 10, 10, &unit_transform(), [0.0, 0.0, 100.0], 0.0, 0.0, None,
        );
        assert!(out.points.is_empty());
        assert!(out.segments.is_empty());
    }

    #[test]
    fn tilted_plane_cuts_flat_dem_along_a_line() {
        let dem = flat_dem(10, 10, 0.0);
        let out = intersect_plane_dem(
            &dem, 10, 10, &unit_transform(), [5.0, 5.0, 0.0], 90.0, 45.0, None,
        );
        assert!(!out.points.is_empty());
        let n = plane_normal(90.0, 45.0);
        for p in &out.points {
            let d = n[0] * (p[0] - 5.0) + n[1] * (p[1] - 5.0) + n[2] * (p[2] - 0.0);
            assert!(d.abs() < 1e-9, "point off plane: {}", d);
        }
    }

    #[test]
    fn nodata_cells_are_skipped() {
        let mut dem = flat_dem(10, 10, 0.0);
        for v in dem.iter_mut() {
            *v = -9999.0;
        }
        let out = intersect_plane_dem(
            &dem, 10, 10, &unit_transform(), [5.0, 5.0, 0.0], 90.0, 45.0, Some(-9999.0),
        );
        assert!(out.points.is_empty());
    }

    #[test]
    fn segments_index_existing_vertices() {
        let dem = flat_dem(10, 10, 0.0);
        let out = intersect_plane_dem(
            &dem, 10, 10, &unit_transform(), [5.0, 5.0, 0.0], 90.0, 45.0, None,
        );
        for s in &out.segments {
            assert!(s[0] < out.points.len() && s[1] < out.points.len());
        }
    }
}
